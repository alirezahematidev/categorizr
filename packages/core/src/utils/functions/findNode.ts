import { MaybeNode, TreeNode } from "$core/types";

export function findNode<T extends TreeNode, const Id extends T["id"]>(tree: readonly T[], id: Id): MaybeNode<T> {
  for (const node of tree) {
    if (node.id === id) return node;

    const foundNode = findNode<T, Id>(node.children as T[], id);

    if (foundNode) return foundNode;
  }
  return undefined;
}
