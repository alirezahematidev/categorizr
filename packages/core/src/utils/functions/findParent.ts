import { MaybeNode, TreeNode } from "$core/types";
import { noChildren } from "../helpers";

export function findParent<T extends TreeNode>(tree: readonly T[], node: T): MaybeNode<T> {
  for (const branch of tree) {
    if (noChildren(branch)) continue;

    if (branch.children.some((child) => child.id === node.id)) return branch;

    const parent = findParent<T>(branch.children as T[], node);

    if (parent) return parent;
  }

  return undefined;
}
