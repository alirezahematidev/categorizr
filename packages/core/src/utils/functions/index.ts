export * from "./findNode";
export * from "./findParent";
export * from "./insertNode";
export * from "./safeInsertNode";
