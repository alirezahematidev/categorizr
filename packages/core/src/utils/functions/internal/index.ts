export * from "./findNode";
export * from "./findParent";
export * from "./containsNode";
