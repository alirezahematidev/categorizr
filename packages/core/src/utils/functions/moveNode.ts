import { todo } from "../helpers";

function moveNode() {
  todo("NO IMPLEMENTATION");
}

export { moveNode };
