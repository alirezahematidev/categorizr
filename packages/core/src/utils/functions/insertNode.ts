import { DestId, InsertCallback, TreeNode } from "$core/types";
import { clone } from "../helpers";
import { findNode } from "./findNode";

function insertNode<T extends TreeNode, const Id extends T["id"]>(tree: readonly T[], destId: DestId<Id>, node: T, callback: InsertCallback<T>): void;
function insertNode<T extends TreeNode, const Id extends T["id"]>(tree: readonly T[], destId: DestId<Id>, node: T): T[];
function insertNode(tree: readonly any[], destId: DestId<any>, node: any, callback?: InsertCallback<any>) {
  const cloneTree = clone(tree);

  if (destId === null) {
    cloneTree.push(node);

    return cloneTree;
  }

  const parentNode = findNode(cloneTree, destId);

  if (!parentNode) throw new Error("cannot found the parent node within given destId.");

  if (!parentNode.children) parentNode.children = [];

  parentNode.children.push(node);

  if (callback) return void callback(cloneTree);

  return cloneTree;
}

export { insertNode };
