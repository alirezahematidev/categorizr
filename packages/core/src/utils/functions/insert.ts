import { DestId, Callback, TreeNode } from "$core/types";
import { clone, exception } from "../helpers";
import { findNode } from "./internal";

function insert<T extends TreeNode>(tree: readonly T[], destId: DestId, node: T): T[];
function insert<T extends TreeNode>(tree: readonly T[], destId: DestId, node: T, callback: Callback<T>): void;
function insert<T extends TreeNode>(tree: readonly T[], destId: DestId, node: T, callback?: Callback<T>) {
  const cloneTree = clone(tree);

  if (destId === null) {
    cloneTree.push(node);

    return cloneTree;
  }

  const parentNode = findNode(cloneTree, destId);

  if (!parentNode) throw exception("insert", "cannot found the parent node within given destId.");

  if (!parentNode.children) parentNode.children = [];

  parentNode.children.push(node);

  if (callback) return void callback(cloneTree);

  return cloneTree;
}

export { insert };
