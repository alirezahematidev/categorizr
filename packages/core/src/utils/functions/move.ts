import { insert } from "./insert";
import { remove } from "./remove";
import { Callback, DestId, TreeNode } from "$core/types";
import { containsNode, findNode, findParent } from "./internal";
import { exception } from "../helpers";

function move<T extends TreeNode>(tree: readonly T[], sourceId: string, destId: DestId): T[];
function move<T extends TreeNode>(tree: readonly T[], sourceId: string, destId: DestId, callback: Callback<T>): void;
function move<T extends TreeNode>(tree: readonly T[], sourceId: string, destId: DestId, callback?: Callback<T>) {
  const node = findNode(tree, sourceId);

  if (containsNode(tree, sourceId, destId)) throw exception("move", "Cannot move the node into its own descendants.");

  if (!node) throw exception("move", "cannot found the node within given id");

  const parent = findParent(tree, node);

  if (parent && parent.id === destId) {
    if (callback) return void callback(tree as T[]);

    return tree as T[];
  }

  const result = insert(remove(tree, sourceId), destId, node);

  if (callback) return void callback(result);

  return result;
}

export { move };
