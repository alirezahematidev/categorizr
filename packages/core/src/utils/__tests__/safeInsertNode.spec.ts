import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";
import type { TreeNode } from "$core/types";
import { safeInsertNode } from "../functions";

describe("safeInsertNode", async () => {
  beforeAll(() => {
    vi.fn().mockImplementation(safeInsertNode);
  });

  afterAll(() => {
    vi.resetAllMocks();
  });

  const { default: CATEGORIES } = await vi.importActual<{ default: TreeNode[] }>("$core/__mocks__");

  it("throws an error when parent node is not found", () => {
    const emptyTree: TreeNode[] = [];

    const node = {
      id: "10",
      name: "category-3",
      children: [],
    };

    expect(safeInsertNode(emptyTree, "3", node)).toStrictEqual(emptyTree);
  });

  it("returns updated tree including the inserted node at first level of tree within null destId", () => {
    const node = {
      id: "5",
      name: "category-5",
      children: [],
    };

    expect(safeInsertNode(CATEGORIES, null, node)).toStrictEqual([
      {
        id: "1",
        name: "category-1",
        children: [
          {
            id: "3",
            name: "sub-category-1",
            children: [],
          },
        ],
      },
      {
        id: "2",
        name: "category-2",
        children: [
          {
            id: "4",
            name: "sub-category-2",
            children: [],
          },
        ],
      },
      {
        id: "5",
        name: "category-5",
        children: [],
      },
    ]);

    expect(safeInsertNode(CATEGORIES, null, node)).toMatchSnapshot();
  });

  it("returns updated tree including the inserted node", () => {
    const node1 = {
      id: "10",
      name: "category-3",
      children: [],
    };

    const node2 = {
      id: "20",
      name: "category-10",
      children: [],
    };

    expect(safeInsertNode(CATEGORIES, "3", node1)).toMatchSnapshot();

    expect(safeInsertNode(CATEGORIES, "10", node2)).toStrictEqual(CATEGORIES);

    safeInsertNode(CATEGORIES, "3", node2, (newTree) => {
      expect(newTree).toMatchSnapshot();
    });

    safeInsertNode(CATEGORIES, "10", node2, (newTree, error) => {
      expect(newTree).toStrictEqual(CATEGORIES);
      expect(error).toStrictEqual(new Error("cannot found the parent node within given destId."));
    });
  });
});
