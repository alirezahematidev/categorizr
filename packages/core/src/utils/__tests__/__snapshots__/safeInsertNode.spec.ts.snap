// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`safeInsertNode > returns updated tree including the inserted node 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [],
            "id": "10",
            "name": "category-3",
          },
        ],
        "id": "3",
        "name": "sub-category-1",
      },
    ],
    "id": "1",
    "name": "category-1",
  },
  {
    "children": [
      {
        "children": [],
        "id": "4",
        "name": "sub-category-2",
      },
    ],
    "id": "2",
    "name": "category-2",
  },
]
`;

exports[`safeInsertNode > returns updated tree including the inserted node 2`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [],
            "id": "20",
            "name": "category-10",
          },
        ],
        "id": "3",
        "name": "sub-category-1",
      },
    ],
    "id": "1",
    "name": "category-1",
  },
  {
    "children": [
      {
        "children": [],
        "id": "4",
        "name": "sub-category-2",
      },
    ],
    "id": "2",
    "name": "category-2",
  },
]
`;

exports[`safeInsertNode > returns updated tree including the inserted node at first level of tree within null destId 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "id": "3",
        "name": "sub-category-1",
      },
    ],
    "id": "1",
    "name": "category-1",
  },
  {
    "children": [
      {
        "children": [],
        "id": "4",
        "name": "sub-category-2",
      },
    ],
    "id": "2",
    "name": "category-2",
  },
  {
    "children": [],
    "id": "5",
    "name": "category-5",
  },
]
`;
